### PR TITLE
DH: stop setting the private key length arbitrarily

### LINKED ISSUE #12071: Duplicate code in DH_set0_pqg
We have this:
https://github.com/openssl/openssl/blob/591315297ec45ada0d31f057c4f6cff7f572bf3e/crypto/dh/dh_lib.c#L222-L233

Maybe the first check wants to check `dh->q` instead of `q`. `dh->q` can be set by the dh_cache_named_group() call when `q == NULL`. Or the first check can just be removed.

When `DH_get_nid(dh) != NID_undef`, we get here:
https://github.com/openssl/openssl/blob/591315297ec45ada0d31f057c4f6cff7f572bf3e/crypto/dh/dh_key.c#L255-L266

If it's dh->length (N) is 0, it still gets set correctly here:
https://github.com/openssl/openssl/blob/591315297ec45ada0d31f057c4f6cff7f572bf3e/crypto/ffc/ffc_key_generate.c#L29-L30

### PR DESCRIPTION
**Pull request by levitte**
The private key length is supposed to be a user settable parameter.
We do check if it's set or not, and if not, we do apply defaults.

Fixes #12071

### CODE REVIEW

**Comment by t8m** [comment]
Actually 224 is appropriate according to the strength of the 2048 bit DH parameters. The key length should be 2*<security strength> according to the FIPS requirements.

**Comment by romen** [comment]
I think I had a related discussion with @slontis somewhere, but I cannot find it.

> Actually 224 is appropriate according to the strength of the 2048 bit DH parameters. The key length should be 2* according to the FIPS requirements.

Not everyone follows FIPS, they might be subject to different jurisdictions than US or just want to be paranoid.
To me it makes sense that this value should be user settable, and only in FIPS mode you validate if this follows US requirements.

I would also be fine if, if not provided, this value is calculated according to a default, and the default being FIPS. But, this would be an optional feature in my eyes, as I don't think we supported such a default up to 1.1.1.

**Comment by romen** [comment]
Also, does this same discussion apply to DSA?
And if not why not? (maybe we fixed it in DSA after the discussion I cannot find, but missed some DH corner cases?)

It could be for example that this surfaced as part of the work on `test/recipes/20-test_cli_fips.t` where only signatures are tested, so DSA is covered, but not DH!

**Comment by levitte** [comment]
@romen, DSAparams doesn't have a recommend length field, so this bit of DH is simply not relevant there.

**Comment by levitte** [comment]
So @t8m,are you saying that a 224 bit size private key is acceptable?

**Comment by t8m** [comment]
> So @t8m,are you saying that a 224 bit size private key is acceptable?

Yes, for 2048 bit DH prime. I would stay with what we currently have as a default but would override if user explicitly sets the desired key length.

**Comment by levitte** [comment]
FIPS requirements are mentioned. What FIPS requirements are we talking about? FIPS-186 is about DSS, right? For key exchange, it seems that SP800-56A is the document to look at. I wonder if we're collectively confusing the two...

**Comment by t8m** [comment]
> FIPS requirements are mentioned. What FIPS requirements are we talking about? FIPS-186 is about DSS, right? For key exchange, it seems that SP800-56A is the document to look at. I wonder if we're collectively confusing the two...

The private key size requirement for DH comes from https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf section 5.6.1.1.1.

**Comment by levitte** [comment]
Ahhhhh!  I just discovered the following comment in the tests that were failing:

https://github.com/openssl/openssl/blob/a829b735b645516041b55746e013692babd8cd31/test/evp_pkey_provided_test.c#L425-L429

https://github.com/openssl/openssl/blob/a829b735b645516041b55746e013692babd8cd31/test/evp_pkey_provided_test.c#L565-L569

This is of course ok for 5.6.1.1.1, but that hasn't been translated well in code, this value *must* be set in the fromdata calls as well, otherwise the length will get some sort of default, which is going to be either zero (unset, so internal defaults will be used)

However, it seems to me that for 5.6.1.1.2, 224 is completely inappropriate (from SP800-56Ar3, underline mine):

> ### **5.6.1.1.2 Using the FIPS 186-Type FFC Parameter-size Sets**
>
> ... For the key-pair generation methods in Sections 5.6.1.1.3 and 5.6.1.1.4, <ins>the value used for the input parameter _N_ **shall** be len(_q_)</ins>, i.e., the bit length of the domain parameter _q_. ...

**Comment by kroeckx** [comment]
You seem to be fixing #12071

**Comment by levitte** [comment]
> You seem to be fixing #12071

Okie

**Comment by levitte** [comment]
> You seem to be fixing #12071

I added a squash commit that adds a line for this in the commit message

**Comment by t8m** [comment]
@levitte 5.6.1.1.2 applies when you use a key generated for DSA (FIPS 186-x) for DH. Or in other words if you do not use a safe prime and have the q value explicitly generated.

**Comment by levitte** [comment]
> @levitte 5.6.1.1.2 applies when you use a key generated for DSA (FIPS 186-x) for DH. Or in other words if you do not use a safe prime and have the q value explicitly generated.

Ah ok.

Either way, I've amended the test fix to specify the recommended private length in the parameters from `EVP_PKEY_fromdata()`

**Comment by t8m** [comment]
> Note that we now have a value for q too in case of a safe prime, because 5.6.1.1.1 requires us to have it. As far as I know, we didn't have it in 1.1.1 for either built-in groups or when loaded from a file. For a 2048 bit DH key using a safe prime, we would generate a 2048 bit number in 1.1.1 and a 2047 bit number in 3.0

And both these values are wrong because for 2048 bit safe prime 224 bit private key should be generated.

**Comment by kroeckx** [comment]
> And both these values are wrong because for 2048 bit safe prime 224 bit private key should be generated.

Why? That doesn't seem to make sense to me.

**Comment by t8m** [comment]
Please read the thread above in regards to NIST SP800-56A requirements. There is also no security gain by having such long private key in comparison to the 224 bit one.

**Comment by kroeckx** [comment]
> Please read the thread above in regards to NIST SP800-56A requirements. There is also no security gain by having such long private key in comparison to the 224 bit one.

As far as I know, this depends on it being a safe prime. rfc7919
has a minimum requirement of 225 for ffdhe2048.

**Comment by levitte** [comment]
*Meanwhile*, it would be nice if someone reviewed the *current* state of this PR.

**Comment by kroeckx** [comment]
Fixees -> Fixes

**Comment by levitte** [comment]
> Fixees -> Fixes

Fixed :wink:

**Comment by kroeckx** [comment]
Did we previously support reading and writing that field for a PKCS#3 file? I guess at least one way of interpreting things is that DH_set_length() will cause it to be written to the file in PKCS#3 format, but not in X9.42 format.

**Comment by levitte** [comment]
> Did we previously support reading and writing that field for a PKCS#3 file?

Nope.  In 1.1.1 `DH_set_length()` is entirely for other applications to use, we only actually use it on `test/ssltest_old.c`.  The pkeyopt `priv_len` is completely new in 3.0.

File: crypto/dh/dh_lib.c [hunk 1]
```diff
@@ -219,18 +219,6 @@ int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 
     ossl_ffc_params_set0_pqg(&dh->params, p, q, g);
     dh_cache_named_group(dh);
-    if (q != NULL)
```
Comment (slontis) [inline crypto/dh/dh_lib.c:222]: This was here for a reason. Please search for every instance of ->length before making impulsive decisions like this.
e.g see dh_check_priv_key().

Comment (slontis) [reply crypto/dh/dh_lib.c:222]: Ok.
So I guess the code you have deleted was equivalent to bounding the length - even if it was doing it in a dodgy way.

It may still need upper bound checks on the length whenever we change q or the private key maybe??

Comment (levitte) [reply crypto/dh/dh_lib.c:222]: > It may still need upper bound checks on the length whenever we change q or the private key maybe??

This is the wrong place to do so, and a read of SP 800-56Ar3 agrees with me.  It does specify conditions on *N* during key generation, in 5.6.1.13 and 5.6.1.1.4, and I do believe we codify that (possibly partly) here:

https://github.com/openssl/openssl/blob/f4bd510503071e20472032dc9d053810a1937e95/crypto/dh/dh_key.c#L259-L261

For PKCS#3, it seems that we use a default when the length is unset:

https://github.com/openssl/openssl/blob/f4bd510503071e20472032dc9d053810a1937e95/crypto/dh/dh_key.c#L273

That does seem a bit sloppy, considering that [PKCS#3](https://www.teletrust.de/fileadmin/files/oid/oid_pkcs-3v1-4.pdf) says this:

> The central authority may optionally select an integer _l_, the private-value length in bits, that satisfies 2<sup><i>l</i>-1</sup>≤<i>p</i>.

**Comment by levitte** [comment]
This PR has now been split.  The import/export fixes are moved to #13166, leaving the more controversial parts here.

**Comment by levitte** [comment]
#13166 got merged, so I rebased this.  Let's see how the CIs fare...

**Comment by levitte** [comment]
A question emerged when looking at this one more time: does the DH `length` *have* to be set when using safe / approved primes?  Leaving it zero simply means to apply sensible defaults, and those defaults could be made to match the requirements just as well, and if it *is* set by a `DH_set_length()` call, the key generation must check it for errors / incompatible values anyway (I suspect our checks are not sufficient).

In summary, it seems to me like the automated assignment of `dh->length` is misplaced, more or less everywhere, and our DH key generation may be lacking certain controls.

**Comment by kroeckx** [comment]
I expect length to be 0 as long as DH_set_length() is not called,
and that we use some kind of default in that case. Calling
DH_set_length() means the defaults are not good enough for
whatever reason.

One of the problems is that in general we don't even know if it's
a prime. But we now have a list of known safe primes, including those
that are approved for FIPS use. Note that some of the published
primes in various RFCs are not safe primes, but as far as I know,
all the ones in our list are.

I think in general, we're really like to deprecate support for DH
primes that are not a named safe prime.

For the named safe primes, we know what a reasonable value is for
the size of the generated key, and when length is to 0, we can use
that size.

For other DH paramaters that are not a named safe prime, we should
probably be more careful, and use different defaults.

**Comment by levitte** [comment]
@kroeckx, you have very eloquently echoed my concerns.  Thank you.

> For other DH paramaters that are not a named safe prime, we should probably be more careful, and use different defaults.

This is what we currently have:

https://github.com/openssl/openssl/blob/f4bd510503071e20472032dc9d053810a1937e95/crypto/dh/dh_key.c#L273

I doubt we can do better, because this is what the PKCS#3 text says:

> ### **7.1 Private-value generation**
> 
>  An integer _x_, the private value, shall be generated privately and randomly. This integer shall satisfy 0 < <i>x</i> < <i>p</i>−1, unless the central authority specifies a private-value length _l_, in which case the integer shall satisfy 2<sup>_l_-1</sup> ≤ <i>x</i> < 2<sup>_l_</sup>.

(note that _l_ doesn't have an upper limit!...  so even though it would seem natural to not allow it to be bigger than the length in bits of _p_, that text doesn't actually say that it must)

**Comment by h-vetinari** [comment]
Something is off with the way github presents the diff here. It should only be the last commit (a handful of lines), cf. https://github.com/openssl/openssl/compare/master...levitte:fix-dh-length, whereas I see 4 commits and +104/-58 lines.

**Comment by levitte** [comment]
> Something is off with the way github presents the diff here

That's quite odd indeed.  Thanks for notifying us.
I dunno what to do about it, though.  Ideas anyone?

**Comment by mattcaswell** [comment]
> That's quite odd indeed. Thanks for notifying us.
I dunno what to do about it, though. Ideas anyone?

My first thought was that the github mirror had fallen behind the master repo. I've seen that happen before with the symptoms that PRs show commits from master that have already been merged. But that doesn't seem to be the case here. The commits showing here are clearly already merged, and have the same commit hashes as are shown for the main github repo mirror. Weird. Looks like a github bug of some description...

**Comment by slontis** [comment]
So I think the code was like this because:
when q was NULL the length was required in order to set the length of the generated private key. (I only added a precalculated q for the safe primes during 3.0).

I dont see a point in allowing a larger length
It is not allowed in fips.. (See ossl_ffc_generate_private_key() Step(2)). I was trying to pass a N value to this function by setting the length. Now that it is not set then N = qbits which is fine (but I think that is why it was propogated - to indicate the maximum length of the private key - which is what the docs describe the length field as).

**Comment by levitte** [comment]
> So I think the code was like this because:
> when q was NULL the length was required in order to set the length of the generated private key. (I only added a precalculated q for the safe primes during 3.0).

For PKCS#3, this length is related to the length of *p*, not of *q*.  And no, the length is not *required*, the code in `dh_key.c` applies appropriate defaults when it isn't set.

I haven't looked enough at our history to see in what chronological order this development came to be.

> I dont see a point in allowing a larger length

Me neither, but the PKCS#3 wording allows it (it may have been an omission)

> I dont see a point in allowing a larger length
> It is not allowed in fips.. (See ossl_ffc_generate_private_key() Step(2)). I was trying to pass a N value to this function by setting the length. Now that it is not set then N = qbits which is fine (but I think that is why it was propagated - to indicate the maximum length of the private key - which is what the docs describe the length field as).

Yeah, and that could just as well be a default when `dh->length()` isn't set.  See, I can't see that SP800-56A says that *N* has to be stored in the DHparams structure, but it does imply that the length *may* be set, and that's that *N* will be assigned to if it is.

File: crypto/dh/dh_backend.c [hunk 2]
```diff
@@ -16,6 +17,41 @@
...
+
+    if (dh == NULL)
+        return 0;
+    ffc = dh_get0_params(dh);
+    if (ffc == NULL)
+        return 0;
+
+    ret = ossl_ffc_params_fromdata(ffc, params);
+    if (ret)
+        dh_cache_named_group(dh); /* This increments dh->dirt_cnt */
```
Comment (slontis) [inline crypto/dh/dh_backend.c:33]: dirty

File: crypto/dh/dh_key.c [hunk 3]
```diff
@@ -271,6 +271,7 @@ static int generate_key(DH *dh)
             if (dh->params.q == NULL) {
                 /* secret exponent length */
                 l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 1;
+                /* TODO: should l be checked for validity here? */
```
Comment (slontis) [inline crypto/dh/dh_key.c:274]: I think we need to not be adding anymore TODO's now.

**Review by slontis — COMMENTED** [review]
Changes look fine to me..
I think this the dh_length should also be a gen_settable, settable and gettable if we are going to deprecate the low level DH functions.

**Comment by levitte** [comment]
> I think the dh_length should also be a gen_settable, settable and gettable if we are going to deprecate the low level DH functions.

It is already gen_settable and gettable.

As for settable, it depends on how we view `DH_set_length()`...
Is it a function of the same sort as `DH_set0_pqg()` et al?  Then we have `EVP_PKEY_fromdata()` to set input parameters when creating a keyparams EVP_PKEY.
Is it a function more like the ECDH cofactor setting function?  Then yeah, a settable is warranted.

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (levitte) [reply crypto/dh/dh_key.c:274]: In that case, do you have an immediate answer?

File: crypto/dh/dh_backend.c (hunk 2 above)
Comment (levitte) [reply crypto/dh/dh_backend.c:33]: Fixed

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (slontis) [reply crypto/dh/dh_key.c:274]: I think it should do the upper bound check (l < p),
but I wont object to no check.
Maybe add the comment related to PKCS3 here also...

**Comment by slontis** [comment]
> As for settable, it depends on how we view DH_set_length()...

I view it as a function that is getting deprecated, and the effective EVP_PKEY replacement would allow you to set it at any time.

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (levitte) [reply crypto/dh/dh_key.c:274]: Err, this line *is* related to PKCS#3

**Comment by levitte** [comment]
> I view it as a function that is getting deprecated, and the effective EVP_PKEY replacement would allow you to set it at any time.

I'm not satisfied with that answer.  We have chosen to replace things like DH_set0_key() and DH_set0_pqg() etc with EVP_PKEY_fromdata().  So I'll ask again, is DH_set_length() the same type of function as DH_set0_pqg(), or is it a different thing (they both set key parameters)?

**Comment by slontis** [comment]
EVP_PKEY_fromdata() was added to load raw key..
I think this question needs to be raised in an OTC meeting in regards to setting any of the deprecated functions including things like DH_set0_pqg(). It is a design decision.
I wont hold the PR up on the question though.

**Comment by t8m** [comment]
IMO the DH length should be settable because this is just an indication of preference for the private key length.

**Comment by levitte** [comment]
> EVP_PKEY_fromdata() was added to load a key from raw data..

It was actually added to construct any valid EVP_PKEY from raw data.  Basically, it's there to replace this sort of code sequence:

``` C
DH *dh = DH_new();
EVP_PKEY *pkey = EVP_PKEY_new();

DH_set0_key(dh, somenum, someothernum);
DH_set0_pqg(dh, some_p, some_q, some_g);
EVP_PKEY_assign_DH(pkey, dh);
```

> I think this question needs to be raised in an OTC meeting in regards to setting any of the deprecated functions including things like DH_set0_pqg(). It is a design decision.

We actually had that discussion quite a while ago...  a year ago perhaps?  At that time, the choice was that we would not allow manipulation of individual key material items (which includes key parameters).  You want to re-open that question?  Feel free.  Me, I'm continuously concerned that the "design" becomes a moving target, that's not a way to reach a point of release...

> I wont hold the PR up on the question though.

Thanks.

**Comment by levitte** [comment]
> IMO the DH length should be settable because this is just an indication of preference for the private key length.

Thank you for stating an opinion on that.  That helps.

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (slontis) [reply crypto/dh/dh_key.c:274]: Sigh.. 
I was referring to what you wrote before about PKCS3.....
See PKCS3 7.1 Private-value generation
An integer x, the private value, shall be generated privately and randomly. This integer shall satisfy 0 < x < p−1, unless the central authority specifies a private-value length l, in which case the integer shall satisfy 2l-1 ≤ x < 2l.

Comment (levitte) [reply crypto/dh/dh_key.c:274]: Ah, gotcha.  Sorry

Comment (slontis) [reply crypto/dh/dh_key.c:274]: If you don't want to deal with the todo now, remove it and add a issue (which is what was discussed in the last OTC meeting).

**Comment by levitte** [comment]
BTW, the dopic of *this* PR is to stop the arbitrary settings of the length, and to ensure that it's passed on an import / export between libcrypto and providers.  I think it should stay that way.

Whether we want the length to be settable individually or via EVP_PKEY_fromdata(), the work can be done in another PR.

With that said, I'd like to see a final review of this PR as it currently stands.

**Comment by levitte** [comment]
Ping

File: include/crypto/dh.h [hunk 4]
```diff
@@ -24,8 +25,10 @@ void dh_cache_named_group(DH *dh);
 
 FFC_PARAMS *dh_get0_params(DH *dh);
 int dh_get0_nid(const DH *dh);
-int dh_ffc_params_fromdata(DH *dh, const OSSL_PARAM params[]);
+int dh_params_fromdata(DH *dh, const OSSL_PARAM params[]);
```
Comment (mattcaswell) [inline include/crypto/dh.h:28]: Should these things have `ossl_` prefixes?

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (mattcaswell) [reply crypto/dh/dh_key.c:274]: I think we should just have the validity check.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (levitte) [reply crypto/dh/dh_key.c:274]: Ok.  Is the bound check that *l* < len(*p*) (where len() is the length in bits) good enough?  Do note that it's stricter than what PKCS#3 stipulates (which is, actually, nothing), albeit logical.

File: include/crypto/dh.h (hunk 4 above)
Comment (levitte) [reply include/crypto/dh.h:28]: *shrug*
"DH" and "dh" is a pre-existing API prefix, I personally find it quite confusing if we end up with some `dh_blah` and some `ossl_dh_bleh` depending on how long they have been around.
Buf more surprisingly, I thought that there had already been a rename fest PR for DH functions...
Either way, I consider the rename stuff off topic for *this* PR.  It would be silly to just rename one function if we should really rename a whole set.

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (levitte) [reply crypto/dh/dh_key.c:274]: Ping?

Comment (levitte) [reply crypto/dh/dh_key.c:274]: You know, there more I ponder this, the more I think I should just remove the comment.  We might think that PKCS#3 is lacking in this regard, but then so it is.

**Comment by levitte** [comment]
Ping

File: crypto/dh/dh_key.c (hunk 3 above)
Comment (levitte) [reply crypto/dh/dh_key.c]: Ah, no, it's me being sloppy...  I forgot about this:

> The central authority may optionally select an integer _l_, the private-value length in bits, that satisfies 2<sup><i>l</i>-1</sup>≤<i>p</i>.

Comment (levitte) [reply crypto/dh/dh_key.c]: Validity check added

**Review by t8m — APPROVED** [review]

**Comment by levitte** [comment]
Merged

28e1d588f14404d480cc2bd38827ecd587625643 DH: stop setting the private key length arbitrarily
